Lazily process bad ER after fraud proof is submitted

NingLin-P opened the pull request:
close #2538 

This PR implements #2538, it contains changes of:
- Process only the first bad ER `submit_fraud_proof` and lazily process other descendant bad ERs in `submit_bundle`
- Introduce a `PendingSlash` operator status and reject the bundle/staking request of the operator with such status
- Reject the fraud proof that targetting the bad ER that is already reported by a previous fraud proof and is under lazily processing (i.e. the descendant bad ER)
- Reject the `switch_domain` request if the operator has unconfirmed ER in the current domain
    - This is not mentioned in #2538 but is necessary, because the `LatestSubmittedER` that we used to determine the pending slash operator is indexed by `(domain_id, operator_id)` and can only check if the operator has submitted bad ER in the current domain, if the operator switch domain then we can't slash the operator for bad ER it submitted in the previous domain.

### Code contributor checklist:
* [x] I have read, understood and followed [contributing guide](https://github.com/subspace/subspace/blob/main/CONTRIBUTING.md)

NingLin-P committed:
Process only the first bad ER in submit_fraud_proof and lazily process the descendants ER

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Introduce the LatestSubmittedER storage item for tracking the latest er submitted by an operator for a specific domain

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Add checks to reject the bundle/staking request of operator that is pending to slash

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Add check to reject fraud proof that targetting pending to prune bad ER

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Add test cases for the pending to the prune bad ER

Signed-off-by: linning <linningde25@gmail.com>

dariolina commented on crates/pallet-domains/src/staking.rs:
```diff
@@ -621,10 +667,17 @@ pub(crate) fn do_switch_operator_domain<T: Config>(
             Error::OperatorNotRegistered
         );
 
+        // Reject switching domain if there is unconfirmed ER submitted by this operator
```
IIUC to be able to switch the operator has to basically go offline (stop producing bundles, but still stake) for a whole challenge period, and then come online, submit switch extrinsic and go offline again until it's processed?
If so, that is quite subpar. I imagine, ideally, we would accept the switch extrinsic, "freeze" the operator stake on the current domain, so they don't produce new bundles and don't reduce other operators' chances and transfer the stake after their last block is out of challenge period.
This said, it can be handled in a separate issue.

vedhavyas commented on crates/pallet-domains/src/block_tree.rs:
```diff
@@ -458,17 +457,24 @@ fn add_new_receipt_to_block_tree<T: Config>(
     let er_hash = execution_receipt.hash::<DomainHashingFor<T>>();
     let domain_block_number = execution_receipt.domain_block_number;
 
+    ensure!(
```
This check feels unnecessary since the codepath ensure the Node is pruned and this is not exposed outside of this crate. No strong opinion though

NingLin-P replied:
Yeah, it is unnecessary but just try to make it robust in case any accidental change breaks the assumption.

vedhavyas commented on crates/pallet-domains/src/block_tree.rs:
```diff
@@ -398,6 +398,16 @@ pub(crate) fn process_execution_receipt<T: Config>(
+
+    // Update the `LatestSubmittedER` for the operator
+    let key = (domain_id, submitter);
+    match receipt_block_number.cmp(&Pallet::<T>::latest_submitted_er(key)) {
```
Even this match does not seem to be required since the Less case will not happen since the above code does not allow it and its confusing to see such a case here. This could simply be a overwrite in both Equal or Greater case IMHO

NingLin-P replied:
Yeah, it's not necessary just try to handle each case explicitly and make it robust, no strong opinion here will change the `match` to `if`

vedhavyas commented on crates/pallet-domains/src/staking.rs:
```diff
@@ -180,7 +181,11 @@ pub struct Operator<Balance, Share, DomainBlockNumber> {
+    /// not assigned to this field directlt, thus MUST use the `status()` method to query the status
+    /// instead.
+    /// TODO: update the filed to `_status` to avoid accidental access in next network reset
+    status: OperatorStatus<DomainBlockNumber>,
```
Why leave a TODO to do it later. Can't we not update the var name here already ?

NingLin-P replied:
Try to avoid migration code to keep compatible with 3h

vedhavyas commented on crates/pallet-domains/src/staking.rs:
```diff
@@ -189,6 +194,46 @@ pub struct Operator<Balance, Share, DomainBlockNumber> {
+
+#[cfg(test)]
+impl<Balance: Zero, Share: Zero, DomainBlockNumber> Operator<Balance, Share, DomainBlockNumber> {
+    pub(crate) fn dummy(
```
PLease do not leave such conditional compiled code everytwhere. We seem to do this all the time. Instead define the necessary dummy function where it is used instead. Would also appreciate if you remove the other instances that we created for benchmarks too.

NingLin-P replied:
>Instead define the necessary dummy function where it is used instead.

The reason to introduce `fn dummy` in this PR is the `Operator:: status` field is changed to private now, thus we have to define the func here.

>Would also appreciate if you remove the other instances that we created for benchmarks too.

Sure, will do that in the incoming PR for updating benchmark result of `submit_fraud_proof`.

vedhavyas commented on crates/pallet-domains/src/staking.rs:
```diff
@@ -621,10 +667,17 @@ pub(crate) fn do_switch_operator_domain<T: Config>(
+        // Reject switching domain if there is unconfirmed ER submitted by this operator
+        // on the `current_domain_id`
+        ensure!(
+            !LatestSubmittedER::<T>::contains_key((operator.current_domain_id, operator_id)),
```
Can you explain the reason why we do not want to allow switching the operators? If the operator continues to produce the block, then essentially, they will never be able to switch unless they go offline for a good period like daria mentioned.

vedhavyas replied:
Thinking on this further, there are two things we can do.
- Get operators for all unconfirmed ERs, either through Host function or through storage and then slash them immediately. We still prune the bad ers lazily though
- Or completely remove switching feature between domains

@jfrank-summit i'm leaning towards removing this switching the domains since with this change, its no different from operator deregistring, waiting and then registering to new domain unless we have a hard requirement to have it though, then we need to do what I suggested in the first point of slashing all operators immediately so that Lazy slashing wont be necessary

NingLin-P replied:
>Can you explain the reason why we do not want to allow switching the operators?

We are using the `LatestSubmittedER ` to check if an operator is pending to slash, `LatestSubmittedER` is indexed by `(domain_id, operator_id)` we can only check if the operator has submitted bad ER in the current domain but can't know if the operator has submitted bad ER in the previous domain.

I planned to handle the switching domain in a separate PR after this one is merged with what Daria suggested in https://github.com/subspace/subspace/pull/2560#discussion_r1503839500, but if we plan to remove switching domain then we are all good and it will be great.

>then we need to do what I suggested in the first point of slashing all operators immediately so that Lazy slashing wont be necessary

To slash operator immediately we need to iterate at most 14_400 number of ERs, duplicate the operator, and then slash all of them. It will still cause a lot of weight.

0xjeremyfrank replied:
I am ok with removing domain switching for operators.

vedhavyas reviewed (commented):
I understand the changes but left some questions/nits

NingLin-P committed:
Remove the unnecessary LatestSubmittedERFallback error from pallet-domains

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Remove the switch_domain extrinsic from pallet-domains

switch_domain is not supported currently due to incompatible with lazily slashing, this
commit only remove the entry of switch_domain and comment out related tests/benchmark to
avoid invasive change to the rest of the code and also we may support it in the future.

Signed-off-by: linning <linningde25@gmail.com>
